### PR TITLE
Add support to compile Go gc assembly

### PR DESCRIPTION
**Pull request by mikekap**
Gc being the standard go compiler; not to be confused with the garbage collector...

Pretty easy - run `go tool asm` for each .s file & use `go tool pack` to shove them into the same archive.

### CODE REVIEW

File: src/com/facebook/buck/go/GoBuckConfig.java [hunk 1]
```diff
@@ -155,8 +166,10 @@ private Tool getGoTool(
     }
 
     CommandTool.Builder builder = new CommandTool.Builder(new HashedFileTool(toolPath.get()));
-    for (String arg : getFlags(extraFlagsConfigKey)) {
-      builder.addArg(arg);
+    if (!extraFlagsConfigKey.isEmpty()) {
```
Comment (Coneko) [inline src/com/facebook/buck/go/GoBuckConfig.java]: Why is this extra check necessary? If `getFlags` doesn't return an empty list for an empty key maybe it should return an optional, or have a precondition check that the key is not empty.

**Comment by Coneko** [comment]
It seems like it should be possible to split the work into three rules, the compilation rule for go sources, the assembling rule for assembly and the packing rules to put them together.

That way you would get higher caching granularity and parallelism.

File: src/com/facebook/buck/go/GoBuckConfig.java (hunk 1 above)
Comment (mikekap) [inline src/com/facebook/buck/go/GoBuckConfig.java]: This is a private helper to the functions above. The packer doesn't accept any flags, so it doesn't have an extra flags config key. I can make it an optional, but it seemed less verbose to use an empty string.

**Comment by mikekap** [comment]
Go builds generally don't need the sub-package parallelism. Go packages are a single shared private namespace: people generally don't throw too much stuff in a single package. Most public go packages have ~2-3 .s files maximum and generally less than 20 .go files. I realize that parallelism is easy to achieve here, but I don't think there's a need for it. If we wanted sub-package parallelism, it would make a lot more sense to target .go files themselves - I *think* you can compile a single .go file to a single .o file, but I don't know if there's a performance improvement there.

**Comment by Coneko** [comment]
Ok, sounds good.

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Comment by Coneko** [comment]
Hmm, the patch doesn't apply, can you rebase on master?

**Pull request closed without merging**